src/Intervals.php

```diff
@@ -0,0 +1,369 @@
+                $borders[] = array('version' => $interval['end']->getVersion(), 'operator' => $interval['end']->getOperator(), 'side' => 'end');
+            }
+        }
+
+        $opSortOrder = self::$opSortOrder;
+        usort($borders, function ($a, $b) use ($opSortOrder) {
+            $order = version_compare($a['version'], $b['version']);
+            if ($order === 0) {
+                return $opSortOrder[$a['operator']] - $opSortOrder[$b['operator']];
+            }
+
+            return $order;
+        });
+
+        $activeIntervals = 0;
+        $intervals = array();
+        $index = 0;
+        $activationThreshold = $constraint->isConjunctive() ? \count($intervalGroups) : 1;
+        $active = false;
+        foreach ($borders as $border) {
+            if ($border['side'] === 'start') {
+                $activeIntervals++;
+            } else {
+                $activeIntervals--;
+            }
+            if (!$active && $activeIntervals >= $activationThreshold) {
+                $intervals[$index] = array('start' => new Constraint($border['operator'], $border['version']));
+                $active = true;
+            }
+            if ($active && $activeIntervals < $activationThreshold) {
+                $active = false;
+
+                // filter out invalid intervals like > x - <= x, or >= x - < x
+                if (
+                    version_compare($intervals[$index]['start']->getVersion(), $border['version'], '=')
+                    && (
+                        ($intervals[$index]['start']->getOperator() === '>' && $border['operator'] === '<=')
+                        || ($intervals[$index]['start']->getOperator() === '>=' && $border['operator'] === '<')
+                    )
+                ) {
+                    unset($intervals[$index]);
+                } else {
+                    $intervals[$index]['end'] = new Constraint($border['operator'], $border['version']);
+                    $index++;
+
+                    if ($stopOnFirstValidInterval) {
+                        break;
+                    }
+                }
+            }
+        }
+
+        return array('intervals' => $intervals, 'devConstraints' => $dev);
+    }
+
+    /**
+     * @phpstan-return array{'intervals': array<int, array{'start': Constraint, 'end': Constraint}>, 'devConstraints': Constraint[]}
+     */
+    private static function generateSingleConstraintIntervals(Constraint $constraint)
+    {
+        $op = $constraint->getOperator();
+        if (substr($constraint->getVersion(), 0, 4) === 'dev-') {
+            $intervals = array();
+
+            // != dev-foo means any numeric version may match
+            if ($op === '!=') {
+                $intervals[] = array('start' => self::zero(), 'end' => self::positiveInfinity());
+            }
+
+            return array('intervals' => $intervals, 'devConstraints' => array($constraint));
+        }
+
+        if ($op[0] === '>') { // > & >=
```

Comment:
as this doesn't handle dev- versions at all, I think this same PR should introduce exceptions if you try to use branch names together with operators other than !=/== ?

Reply:
Nevermind, I missed that those are all covered above already.